coriolis/providers/replicator.py

```diff
@@ -429,163 +402,133 @@ def _setup_replicator_user(self, ssh):
 
-        utils.exec_ssh_cmd(
-            self._ssh, "sudo cp -f %s /tmp/creds.zip" % zipFile, get_pty=True)
-        utils.exec_ssh_cmd(
-            self._ssh, "sudo chmod +r /tmp/creds.zip", get_pty=True)
+        exist = []
+        for i in (remote_ca_crt, remote_client_crt, remote_client_key,
+                  remote_srv_crt, remote_srv_key):
+            exist.append(utils.test_ssh_path(ssh, i))
 
-        sftp = paramiko.SFTPClient.from_transport(
-            self._ssh.get_transport())
-        try:
-            sftp.get("/tmp/creds.zip", localCertZip, callback=progressf)
-        finally:
-            sftp.close()
+        force_fetch = False
+        if not all(exist):
+            utils.exec_ssh_cmd(
+                ssh, "sudo mkdir -p %s" % remote_base_dir, get_pty=True)
+            utils.exec_ssh_cmd(
+                ssh,
+                "sudo %(replicator_cmd)s gen-certs -output-dir "
+                "%(cert_dir)s -certificate-hosts %(extra_hosts)s" % {
+                    "replicator_cmd": REPLICATOR_PATH,
+                    "cert_dir": remote_base_dir,
+                    "extra_hosts": ip,
+                },
+                get_pty=True)
+            utils.exec_ssh_cmd(
+                ssh, "sudo chown -R %(user)s:%(user)s %(cert_dir)s" % {
+                        "cert_dir": remote_base_dir,
+                        "user": REPLICATOR_USERNAME
+                    }, get_pty=True)
+            force_fetch = True
+
+        exists = []
+        for i in (ca_crt, client_crt, client_key):
+            exists.append(os.path.isfile(i))
+
+        if not all(exists) or force_fetch:
+            # certificates either are missing, or have been regenerated
+            # on the writer worker. We need to fetch them.
+            self._fetch_remote_file(ssh, remote_ca_crt, ca_crt)
+            self._fetch_remote_file(ssh, remote_client_crt, client_crt)
+            self._fetch_remote_file(ssh, remote_client_key, client_key)
 
-        zFile = zipfile.ZipFile(localCertZip)
-        zFile.extractall(path=self._cert_dir)
-        zFile.close()
-        os.remove(localCertZip)
         return {
-            "client_cert": clientCrt,
-            "client_key": clientKey,
-            "ca_cert": caCert,
+            "local": {
+                "client_cert": client_crt,
+                "client_key": client_key,
+                "ca_cert": ca_crt,
+            },
+            "remote": {
+                "srv_crt": remote_srv_crt,
+                "srv_key": remote_srv_key,
+                "ca_crt": remote_ca_crt,
+            },
         }
 
+    @utils.retry_on_error()
+    def _setup_replicator(self, ssh):
+        # copy the binary, set up the service, generate certificates,
+        # start service
+        state_file = self._get_replicator_state_file()
+        self._copy_file(ssh, state_file, REPLICATOR_STATE)
+        utils.exec_ssh_cmd(
+            ssh, "sudo chmod 755 %s" % REPLICATOR_STATE, get_pty=True)
```

Comment:
If adding `get_pty=True` to `exec_ssh_cmd` is so common, maybe we should consider setting `True` as the default value for the kwarg.

Reply:
get_pty is only needed if we run ```sudo``` on systems where it was explicitly configured to need a real ```pty```. Changing this default is outside the scope of this PR though.